gtsam/nonlinear/utilities.h

```diff
@@ -142,6 +143,22 @@ Matrix extractPose3(const Values& values) {
   return result;
 }
 
+/// Extract all Vector values into a single tall vector
```

Comment:
A single tall vector? I don’t like that at all :-) at least Pose3 gets extracted into a matrix.
What is the use case here, really?
And what happens if you want fixed size vectors?

Reply:
This cannot handle fixed size vectors (they don't get returned), but from matlab it's actually quite difficult (impossible other than Point2/Point3?) to place fixed-size vectors into a Values object anyway in the current develop branch.

The output is a single tall vector since (a) it's very rare to have fixed-size vectors anyway and (b) since the vectors are potentially different sizes, then you can't place in a matrix without zero-padding.  It's actually quite easy in matlab to `reshape(vectors, N, [])` if you know all your vectors are the same size.

As for the use case, lots of linear dynamics problems will have vectors (as we have seen) and extracting the vectors from a Values result using `for i =1:N; result.atVector(i); end` is often glacially slow.  I remember actually having this experience over the summer as well and as a dirty hack I even manually parsed the serialized strings to make extracting values faster in matlab (Adam was on an old version of gtsam so I couldn't make PR's to the develop branch).  Therefore, I believe these types of convenience extraction functions have utility.

The choice to use dynamic vectors instead of fixed-size is less obvious, but my naive and subjective opinion is that for the flexibility it is reasonable.

Reply:
Maybe it's better to figure out why the extraction is so slow. Recently Varun has revamped a lot of the old code, so maybe performance has improved? (and the code is now easier to read)

Reply:
Hi All,

I cannot tell if this issue has been resolved or not. @gchenfc I am still using the custom extract with the reshape command, but I know that @dellaert wanted another solution. What is the end user (like a corporate sponsor) suppose to do to use the extract for vectors from a value object in MATLAB?

Reply:
You do a `values.atVector(key)`. @mattking-smith please do not approve PRs without providing _sufficient reasoning_ on the approval especially if your issue is not quite as related. :slightly_smiling_face:

Reply:
> You do a `values.atVector(key)`. @mattking-smith please do not approve PRs without providing _sufficient reasoning_ on the approval especially if your issue is not quite as related. 🙂

Also, Matt, can't approve your *own* PR :-)

Reply:
@varunagrawal I think this was an efficiency question.  I probably should have made the PR description more clear, but to clarify:

Doing `values.atVector(key)` for, say, a hundred keys is glacially slow and will actually consume the vast majority of the runtime for many problems.  I was telling Matt that one time I actually used something like `fscanf('%f ', values.serialize())` or something because that was a thousand times faster than `values.atVector(key)` for every key.

I'm pretty sure the other functions in `utilities.h` were written for the same reason - someone else realized how slow it is to loop through keys in matlab.

I can work together with matt to fix this up after ICRA.

Reply:
Yes, it's a valid PR. I only have issues with the API and the "one big vector". See my review comment.

Reply:
My response was more directed towards the corporate sponsor part. I imagine for most cases, getting all the vectors as one big matrix would not be a major necessity.